src/viam/sdk/common/proto_type.hpp

```diff
@@ -27,64 +27,82 @@ struct get_helper {
-    template <typename V>
-    static std::shared_ptr<const AttributeMap::element_type> get(const V& v) {
-        auto* const result = boost::get<AttributeMap>(&v);
-        return result ? *result : nullptr;
-    }
-};
-
-}  // namespace prototype_details
+}  // namespace value_type_details
 
-class ProtoType {
+class AttributeMap {
    public:
-    ProtoType() {
-        proto_type_ = boost::blank();
-    }
-
-    explicit ProtoType(bool b) : proto_type_(std::move(b)) {}
-    explicit ProtoType(std::string s) : proto_type_(std::move(s)) {}
-    explicit ProtoType(const char* c) : proto_type_(std::string(c)) {}
-    explicit ProtoType(int i) : proto_type_(std::move(i)) {}
-    explicit ProtoType(double d) : proto_type_(std::move(d)) {}
-    explicit ProtoType(AttributeMap m) : proto_type_(std::move(m)) {}
-    explicit ProtoType(std::vector<std::shared_ptr<ProtoType>> v) : proto_type_(std::move(v)) {}
-    explicit ProtoType(const google::protobuf::Value& value);
-
-    google::protobuf::Value proto_value();
-    friend bool operator==(const ProtoType& lhs, const ProtoType& rhs);
-
-    template <typename T>
-    auto get() {
-        return prototype_details::get_helper<T>::get(proto_type_);
-    }
-
-    template <typename T>
-    auto get() const {
-        return prototype_details::get_helper<T>::get(proto_type_);
-    }
+    class value_type;
+
+    using map_type = std::unordered_map<std::string, value_type>;
+
+    class value_type {
+       public:
+        template <typename T>
+        auto get() {
+            return value_type_details::get_helper<T>::get(value_);
+        }
+
+        template <typename T>
+        auto get() const {
+            return value_type_details::get_helper<T>::get(value_);
+        }
+
+        using base_value_types_ =
+            boost::variant<boost::blank, std::string, const char*, int, float, double, bool>;
+        using value_types = boost::make_recursive_variant<
+            base_value_types_,
+            std::unordered_map<std::string, boost::recursive_variant_>,
+            std::vector<boost::recursive_variant_>>::type;
+
+        value_type();
+        value_type(std::string s);
+        value_type(const char* c);
+        value_type(int i);
+        value_type(float f);
+        value_type(double d);
+        value_type(bool b);
+        value_type(std::vector<value_type> v);
+        value_type(map_type);
+        value_type(google::protobuf::Value value);
+
+        google::protobuf::Value to_proto() const;
+
+       private:
+        value_types value_;
+    };
+
+    AttributeMap();
+    ~AttributeMap();
+    // CR erodkin: add comments to all of this probably
+    google::protobuf::Struct to_proto() const;
+
+    bool contains(const std::string& key) const;
+
+    // CR erodkin: consider how to return a ref or pointer to the value type such that we don't have
+    // to copy the `value_type`, while retaining guarantees of const-ness.
+    value_type get(const std::string& key) const;
```

Comment:
What `value_type` gets returned if the item is not found? Or does it throw?

Reply:
Yeah good question. My intuition is returning a `boost::blank` as the default is best, people can check for the "not found" case pretty easily (`if (!foo)`) and it saves having to potentially large chunks of code in a `try/catch`.